### ISSUE TITLE
[BUG - Release] Unable to hook Blueprint Library functions

### ISSUE THREAD

**Issue by Okaetsu** [UE4SS-RE/RE-UE4SS#467]
Hopefully not a duplicate, but from a quick look I couldn't find the issue being mentioned anywhere.

**Branch or Release**
3.0.1 Release

**Game and Engine Version**
Palworld (UE 5.1.1), Stray (UE 4.27) and most likely every other game.

**Describe the bug**
When using Lua or LiveView to hook to any blueprint library functions, it freezes the game and returns an error:
```
[04:29:32] [RegisterHook] Registered native hook (5, 6) for Function /Script/Engine.KismetMathLibrary:Add_IntInt
[04:29:32] Tried to execute UFunction::FuncPtr hook but there was no function map entry for UFunction:
	0000000000000000 None.
	No original function to execute.
	Executing ProcessInternal instead.
```

**To Reproduce**
1. Launch game
2. Hook any blueprint library function
3. Wait for the blueprint library function to run or call a blueprint library function manually
4. Game is now frozen with an error in the console until you restart the game

**Expected behavior**
Same as other normally hooked functions, no crashing.

**Comment by UE4SS** [comment]
I'm unable to reproduce this in latest Palworld.
```lua
RegisterHook("/Script/Engine.KismetMathLibrary:Add_IntInt", function()
	print("/Script/Engine.KismetMathLibrary:Add_IntInt\n")
end)
RegisterKeyBind(Key.I, function()
	local Math = StaticFindObject("/Script/Engine.Default__KismetMathLibrary")
	local Val = Math:Add_IntInt(2, 3)
	print(string.format("Val: %s\n", Val))
end)
```
Result:
![2024-04-13_15-36_add_int_int_success](https://github.com/UE4SS-RE/RE-UE4SS/assets/91646798/0f45f993-bfd2-4831-a763-e69e4d5e51e9)

**Comment by UE4SS** [comment]
For some reason, my own call doesn't trigger this problem but using the menu does.

**Comment by Okaetsu** [comment]
> I'm unable to reproduce this in latest Palworld.

Try hooking `Function /Script/Pal.PalUtility:AdjustActorToFloor` and load into a World, should immediately give an error once the load is finished.

**Comment by glgen** [comment]
This happens when hooked native UFunction is called from blueprint via EX_CallMath kismet expression.
UE4SS uses Stack.CurrentNativeFunction to find out what function was called. But in EX_CallMath realization in UE source the line setting Stack.CurrentNativeFunction to called function is commented out, so hook can't find original function.

Also, when native function is called from any kismet bytecodes, it doesn’t seem to put params in Stack.Locals, which ue4ss uses for parameters values, and they becomes random values.

**Comment by UE4SS** [comment]
Yeah this is a known long-standing problem.
I'm unsure how to fix it.

**Comment by UE4SS** [comment]
@Okaetsu If I make a draft PR, can you, when you have time, hook a bunch of different functions and see if you can manage to cause any problems ?

**Comment by Okaetsu** [comment]
> @Okaetsu If I make a draft PR, can you, when you have time, hook a bunch of different functions and see if you can manage to cause any problems ?

Yeah

**Comment by UE4SS** [comment]
> > @Okaetsu If I make a draft PR, can you, when you have time, hook a bunch of different functions and see if you can manage to cause any problems ?
> 
> Yeah

Actually it involves changing both UE4SS and the Unreal module.
I'll put the patches here and you'll need to apply them locally instead, that'll probably be the easiest way to do this.
I hate submodules.
These are based on main, not your bp_hooks, hopefully that won't cause too much of a headache for you.

UE4SS:
```patch
Index: UE4SS/src/Mod/LuaMod.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/UE4SS/src/Mod/LuaMod.cpp b/UE4SS/src/Mod/LuaMod.cpp
--- a/UE4SS/src/Mod/LuaMod.cpp	(revision 7d6f790dc0a027935cb14da67d12ab42c345015e)
+++ b/UE4SS/src/Mod/LuaMod.cpp	(date 1713023589290)
@@ -127,12 +127,13 @@
         LuaType::RemoteUnrealParam::construct(lua_data.lua, &context.Context, s_object_property_name);
 
         // Attempt at dynamically fetching the params
-        uint16_t return_value_offset = context.TheStack.CurrentNativeFunction()->GetReturnValueOffset();
+        const auto FunctionBeingExecuted = lua_data.unreal_function;
+        uint16_t return_value_offset = FunctionBeingExecuted->GetReturnValueOffset();
 
         // 'ReturnValueOffset' is 0xFFFF if the UFunction return type is void
         lua_data.has_return_value = return_value_offset != 0xFFFF;
 
-        uint8_t num_unreal_params = context.TheStack.CurrentNativeFunction()->GetNumParms();
+        uint8_t num_unreal_params = FunctionBeingExecuted->GetNumParms();
         if (lua_data.has_return_value)
         {
             // Subtract one from the number of params if there's a return value
@@ -145,7 +146,7 @@
         {
             // int32_t current_param_offset{};
 
-            for (Unreal::FProperty* func_prop : context.TheStack.CurrentNativeFunction()->ForEachProperty())
+            for (Unreal::FProperty* func_prop : FunctionBeingExecuted->ForEachProperty())
             {
                 // Skip this property if it's not a parameter
                 if (!func_prop->HasAnyPropertyFlags(Unreal::EPropertyFlags::CPF_Parm))
@@ -276,12 +277,13 @@
             LuaType::RemoteUnrealParam::construct(lua_data.lua, &context.Context, s_object_property_name);
 
             // Attempt at dynamically fetching the params
-            uint16_t return_value_offset = context.TheStack.CurrentNativeFunction()->GetReturnValueOffset();
+            const auto FunctionBeingExecuted = lua_data.unreal_function;
+            uint16_t return_value_offset = FunctionBeingExecuted->GetReturnValueOffset();
 
             // 'ReturnValueOffset' is 0xFFFF if the UFunction return type is void
             lua_data.has_return_value = return_value_offset != 0xFFFF;
 
-            uint8_t num_unreal_params = context.TheStack.CurrentNativeFunction()->GetNumParms();
+            uint8_t num_unreal_params = FunctionBeingExecuted->GetNumParms();
             if (lua_data.has_return_value)
             {
                 // Subtract one from the number of params if there's a return value
@@ -289,7 +291,7 @@
                 --num_unreal_params;
 
                 // Set up the return value param so that Lua can access the original return value
-                auto return_property = context.TheStack.CurrentNativeFunction()->GetReturnProperty();
+                auto return_property = FunctionBeingExecuted->GetReturnProperty();
                 auto return_property_type = return_property->GetClass().GetFName();
                 int32_t name_comparison_index = return_property_type.GetComparisonIndex();
                 if (LuaType::StaticState::m_property_value_pushers.contains(name_comparison_index))
@@ -306,7 +308,7 @@
             bool has_properties_to_process = lua_data.has_return_value || num_unreal_params > 0;
             if (has_properties_to_process && context.TheStack.Locals())
             {
-                for (Unreal::FProperty* func_prop : context.TheStack.CurrentNativeFunction()->ForEachProperty())
+                for (Unreal::FProperty* func_prop : FunctionBeingExecuted->ForEachProperty())
                 {
                     // Skip this property if it's not a parameter
                     if (!func_prop->HasAnyPropertyFlags(Unreal::EPropertyFlags::CPF_Parm))

```

Unreal:
```patch
Index: src/UFunctionStructs.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/UFunctionStructs.cpp b/src/UFunctionStructs.cpp
--- a/src/UFunctionStructs.cpp	(revision d3d482dc8d55bcccca0426b1e231562a4256f0b4)
+++ b/src/UFunctionStructs.cpp	(date 1713023397096)
@@ -137,7 +137,8 @@
         try
         {
             HookedUFunctionMap& FunctionMap = GetHookedFunctionsMap();
-            const auto Iterator = FunctionMap.find(TheStack.CurrentNativeFunction());
+            const auto FunctionBeingExecuted = TheStack.Code() ? *std::bit_cast<UFunction**>(&TheStack.Code()[0-sizeof(uint64)]) : TheStack.CurrentNativeFunction();
+            const auto Iterator = FunctionMap.find(FunctionBeingExecuted);
 
             if (Iterator == FunctionMap.end())
             {
@@ -145,18 +146,18 @@
                 // This could be due to a bug.
                 // Either way, if CurrentNativeFunction is nullptr, it might be safe to call ProcessInternal.
                 // If CurrentNativeFunction is non-nullptr, then we call whatever is stored in CurrentNativeFunction.Func.
-                if (TheStack.CurrentNativeFunction())
+                if (FunctionBeingExecuted)
                 {
                     Output::send<LogLevel::Warning>(STR("Tried to execute UFunction::FuncPtr hook but there was no function map entry for UFunction:\n\t{:016X} {}.\n\tExecuting original function instead.\n"),
-                                                    std::bit_cast<uintptr_t>(TheStack.CurrentNativeFunction()),
-                                                    TheStack.CurrentNativeFunction() ? TheStack.CurrentNativeFunction()->GetFullName() : STR("None"));
-                    TheStack.CurrentNativeFunction()->GetFuncPtr()(Context, TheStack, RESULT_DECL);
+                                                    std::bit_cast<uintptr_t>(FunctionBeingExecuted),
+                                                    FunctionBeingExecuted ? FunctionBeingExecuted->GetFullName() : STR("None"));
+                    FunctionBeingExecuted->GetFuncPtr()(Context, TheStack, RESULT_DECL);
                 }
                 else
                 {
                     Output::send<LogLevel::Warning>(STR("Tried to execute UFunction::FuncPtr hook but there was no function map entry for UFunction:\n\t{:016X} {}.\n\tNo original function to execute.\n\tExecuting ProcessInternal instead.\n"),
-                                                    std::bit_cast<uintptr_t>(TheStack.CurrentNativeFunction()),
-                                                    TheStack.CurrentNativeFunction() ? TheStack.CurrentNativeFunction()->GetFullName() : STR("None"));
+                                                    std::bit_cast<uintptr_t>(FunctionBeingExecuted),
+                                                    FunctionBeingExecuted ? FunctionBeingExecuted->GetFullName() : STR("None"));
                     static auto Object_ExecuteUbergraph = UObjectGlobals::StaticFindObject<UFunction*>(nullptr, nullptr, STR("/Script/CoreUObject.Object:ExecuteUbergraph"));
                     static auto ProcessInternal = Object_ExecuteUbergraph->GetFuncPtr();
                     Output::send(STR("Object_ExecuteUbergraph: {}\n"), (void*)Object_ExecuteUbergraph);
@@ -176,7 +177,7 @@
                 }
                 catch (std::exception& e)
                 {
-                    Output::send(STR("Error executing hook pre-callback {}: {}\n"), TheStack.CurrentNativeFunction()->GetPathName(), to_wstring(e.what()));
+                    Output::send(STR("Error executing hook pre-callback {}: {}\n"), FunctionBeingExecuted->GetPathName(), to_wstring(e.what()));
                 }
 
                 Iterator->second.GetOriginalFuncPtr()(Context, TheStack, RESULT_DECL);
@@ -187,7 +188,7 @@
                 }
                 catch (std::exception& e)
                 {
-                    Output::send(STR("Error executing hook post-callback {}: {}\n"), TheStack.CurrentNativeFunction()->GetPathName(), to_wstring(e.what()));
+                    Output::send(STR("Error executing hook post-callback {}: {}\n"), FunctionBeingExecuted->GetPathName(), to_wstring(e.what()));
                 }
 
                 Iterator->second.bIsMidExecution = false;
@@ -195,7 +196,7 @@
         }
         catch (std::exception& e)
         {
-            Output::send(STR("Error executing hooked function {}: {}\n"), TheStack.CurrentNativeFunction()->GetPathName(), to_wstring(e.what()));
+            Output::send(STR("Error executing hooked function: {}\n"), to_wstring(e.what()));
         }
     }
 }
```

**Comment by glgen** [comment]
As i see all other bytecodes calls UFunction::Invoke (at least when function isn't remote) which sets Stack.CurrentNativeFunction, so maybe we don't need Stack.Code for other bytecodes

EX_VirtualFunction and EX_FinalFunction -> UObject::CallFunction -> UFunction::Invoke (when not remote)
EX_LocalVirtualFunction and EX_LocalFinalFunction -> ProcessLocalFunction -> UFunction::Invoke

Also, EX_FinalFunction and EX_LocalFinalFunction stores function pointer, but EX_VirtualFunction and EX_LocalVirtualFunction stores function name.

**Comment by Okaetsu** [comment]
Did some quick testing and while the problematic hooked functions don't freeze the game anymore, the values provided within the callback seem to be filled with junk except for Context which works normally. 

LiveView doesn't display any data whenever those functions are called and as for Lua, when I tried accessing int values from `/Script/Engine.KismetMathLibrary:Add_IntInt`, the second parameter seemed to work fine while the others were filled with random data. This happens with other blueprint library functions aswell.

**Comment by UE4SS** [comment]
> Did some quick testing and while the problematic hooked functions don't freeze the game anymore, the values provided within the callback seem to be filled with junk except for Context which works normally.
> 
> LiveView doesn't display any data whenever those functions are called and as for Lua, when I tried accessing int values from `/Script/Engine.KismetMathLibrary:Add_IntInt`, the second parameter seemed to work fine while the others were filled with random data. This happens with other blueprint library functions aswell.

As far as I can tell, BP function params are either inlined into the byte-code or in `Stack.Locals`.
This would explain why the second parameter works but not the first one.
The second one is probably a local, which UE4SS handles, while the first one is probably inlined which UE4SS doesn't handle.
Not sure yet how to fix this.
The byte-code in this screenshot hopefully explains the problem:
![2024-04-13_23-11_ue_bytecode_callmath](https://github.com/UE4SS-RE/RE-UE4SS/assets/91646798/962c2f0a-10ed-4721-8783-0204eee0cf34)

**Comment by UE4SS** [comment]
We would probably need to be able to actually to step backwards into the byte-code until the param type is found (i.e. `EX_LocalVariable` or `Ex_IntConst`) and once determined, we'd step forward and retrieve the value appropriately.
There's maybe a possibility of reusing some of the code in the KismetDebugger mod for this.

**Comment by UE4SS** [comment]
If the original crash is going to be fixed at any point soon, we'll have to delay the param fix for later.
A new issue can be created for it.
